### DISCUSSION TITLE
Have anyone got the production SecOnion works in GCP with Packet Mirroring yet?

### DISCUSSION THREAD

**Discussion by coolslaz** [Security-Onion-Solutions/securityonion#2171] [Unsupported Versions]
Just wondering have anyone here got the production SecOnion works in the GCP with Packet Mirroring yet?

**Comment by weslambert** [comment]
Hi @coolarnie , I'll be developing some documentation to assist with this soon.

**Reply by Stephen-Ridgway** [reply]
Hi @coolarnie and @weslambert , did this documentation get written? I've got a production GCP sensor to rebuild from 16.04 to v2 and I'm struggling to get networking set up.

**Reply by coolslaz** [reply]
@Stephen-Ridgway I don't think so. I've been waiting on this documentation as well.

**Reply by Stephen-Ridgway** [reply]
In ubuntu 16.04 I used a bridge interface as the management interface and a single ethernet interface as the sniffing interface. It worked well, but I can't get the same architecture to work on either Ubuntu 18.04 or Centos7

If I create the bridge interface on Ubuntu, SO doesn't recognise it during the setup (actually doesn't recognise any interfaces). I can't even get the bridge to work at all in Centos7, which I guess is a GCP thing, as I've done it before on physical Centos servers.

**Reply by coolslaz** [reply]
I used to use the CloudClient with ubuntu 16.04 and openVPN back into our SecOnion which is used to hosted on premise.
Ever since we've migrated into Cloud, we haven't get any chance to have the fully production SecOnion in GCP yet.

For what I know, we'll need to use GCP Packet Mirroring to get SO to works which is much more complex setup if compared with AWS. 
- We'll need to setup Load Balancer
- We'll need to setup instance group
- We'll need to setup Packet Mirroring policy

Reference:
How Do I Use - GCP Packet Mirroring - https://www.youtube.com/watch?v=Cd4N9nRhOkU
GCP Packet Mirroring - From Zero to Pcap - https://www.youtube.com/watch?v=Cd4N9nRhOkU

At the end, once setup done in GCP end. We still need to get SO part works which is where I really need some help or @weslambert SecOnion deployment in GCP Packet Mirroring documentation. We're relying on [terraform deployment.](https://github.com/Security-Onion-Solutions/securityonion-cloud/tree/master/terraform/gcp)

**Reply by weslambert** [reply]
I'll try to update the documentation this week, as time allows.

**Reply by cyberkimh** [reply]
Thanks @weslambert I'd be really interested in seeing this too.

**Reply by coolslaz** [reply]
Hiya @weslambert Any chance the doco will be ready soon? Thanks again.

**Reply by Stephen-Ridgway** [reply]
I'm in a world of pain with GCP :(
@weslambert, any ETA on the documentation? Really appreciate your help with this!

**Reply by coolslaz** [reply]
@Stephen-Ridgway +1 I'm awaiting for long to have the Cloud SOC PoC with SecOnion established within the GCP to monitor our traffic as well.

**Reply by weslambert** [reply]
I'll update the docs as soon as I can, by no defined date at the moment -- I've set it up several times, just need to get it written down.

**Comment by Stephen-Ridgway** [comment]
So here's what I've discovered so far in my travels in GCP land:

Basics:

- GCP will allow only one interface in any one VPC. As VPCs are global (unlike AWS), this is rather restrictive. I've had to set up a management VPC in addition to the VPC I want to mirror traffic from
- The order of the network interfaces is important. GCP insists on routing the default gateway via the first interface (eth0, in my case). Yes, it's possible to force the default gw via another interface, but that creates a world of pain. Eth0 needs, therefore, to be in the management VPC.
- It's not possible to reliably change the hostname after a GCP instance is built. The normal things, like changing /etc/hostname, get's overwritten at reboot. This matters because of the way the sensor and the master communicate (salt keys etc).

Lessons learnt:

- Centos must be up to date and the latest SO pulled via git BEFORE starting the installation. I've had loads of problems with dependencies and versioning of components (especially python). If all else fails or things have got screwed up by previous attempts at installation, running 'yum clean all' resets the repos and seems to help (not sure if this is obvious to everyone else, but coming from a Debian background, I'm still getting to grips with Centos. I stopped using RedHat in 2003...)
- Make sure you have a means to get back onto the sensor once the installation is finished. There was a warning about changing ssh keys that I skimmed just before reboot, then found that GCP was no longer able to push my default ssh key to the instance. It's possible to fix this by either allowing OS pwd logons before starting the setup or by manually copying your own public key to authorised_hosts in you .ssh directory. GCP Cloud Console will allow you on to a box that you've locked yourself out of, but it's slow and painful.
- Any mismatch between resolved names on the master and sensor seems to terminally break setup. It's worth double-checking that the right name resolves to both 127.0.0.1 and the correct IP address of the correct interface before running setup.
- Network settings don't always 'stick'. I'm still confused about how network settings are configured. I rue the day that ifup/down was replaced with NetworkManager/nmcli/networkd etc etc. GCP seems to make changes on reboot, but I can't see why.

Remaining issues:

- the network bond setup doesn't get configured. Neither bond0 nor bond0-slave-ethx get set up. I'm currently trying to work out where in the setup script this is configured so I can troubleshoot what's gone wrong. Eth1 (my sniffing interface) gets correctly configured as a promic interface, but keeps it's IP address (can't find where that's set. Dhcp is set to 'no' and the interface is not managed by NetworkManager...)

Hope this helps someone (it was cathartic to write it down, anyway) :)

**Reply by weslambert** [reply]
Thanks for your feedback.  As you stated, you'll want to ensure the box is updated before cloning and running setup.  Also, we don't use a bond configuration for cloud instances of Security Onion 2, as several providers don't allow it.  For network interfaces, you will indeed need to have them in different networks/subnets, however, this isn't too bad from a best practices perspective 😄 .  As far as the internal IP address assigned to the promisc interface, is it causing issues for you?  I'm pretty sure this is necessary for GCP to be able to route the mirrored traffic accordingly.  I'll get the process documented very soon.

**Reply by Stephen-Ridgway** [reply]
Thanks @weslambert !! I've only just got to the point of setting up the mirrored traffic, so I'll test it and report back :)

**Comment by Stephen-Ridgway** [comment]
Got there.

https://stephen-ridgway.medium.com/how-to-build-a-security-onion-sensor-in-gcp-8540f3ce0dcd
